community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -472,6 +474,98 @@ def update_topology(lkp: util.Lookup) -> None:
         log.debug("Topology configuration updated. Reconfiguring Slurm.")
         util.scontrol_reconfigure(lkp)
 
+
+def delete_reservations(reservation_map: Dict[str, str]) -> None:
+    slurm_delete_reservation = "scontrol delete reservation {}"
+    for reservation_name in list(reservation_map.keys()):
+      util.run(slurm_delete_reservation.format(reservation_name), timeout=30)
+
+
+def update_slurm_reservation_maintenance(node: str, starttime: str,
+                                         reservation_map: Dict[str, Tuple[str, str]]) -> None:
+    slurm_create_reservation = "scontrol create reservation user=root starttime={} duration=180 nodes={} reservationname={}"
+    slurm_delete_reservation = "scontrol delete reservation {}"
+
+    reservation_name = node + "_maintenance"
+    ## If reservation exists for maintenance, update reservation if time window
+    ## for scheduled maintenance changes.
+    if reservation_name in list(reservation_map.keys()):
+      nodes, maintenance_start_time = reservation_map[reservation_name]
+      del reservation_map[reservation_name]  # remove the reservation from the map.
+
+      if maintenance_start_time == starttime:
+        return
+
+      util.run(slurm_delete_reservation.format(reservation_name), timeout=30)
+      util.run(slurm_create_reservation.format(starttime, nodes, reservation_name), timeout=30)
+
+    else: # Reservation doesn't exist, make reservation during maintenance window.
+      util.run(slurm_create_reservation.format(starttime, node, reservation_name), timeout=30)
+
+
+def get_slurm_reservation_maintenance() -> Dict[str, Tuple[str, str]]:
+    slurm_show_reservation = "scontrol show reservation --json"
+    res = util.run(slurm_show_reservation, timeout=30)
+    all_reservations = json.loads(res.stdout)
+    reservation_map = {}
+
+    for reservation in all_reservations['reservations']:
+        reservation_name = reservation['name']
+        nodes = reservation['node_list']
+        linux_time = reservation['start_time']['number']
+
+        if reservation_name is None or nodes is None or linux_time is None:
+          continue
+
+        # Check if the reservation is for scheduled maintenance.
+        maintenance_reservation = nodes + "_maintenance"
+        if reservation_name != maintenance_reservation:
+          continue
+
+        # Update maintenance start time compatible with slurm reservation format.
+        start_time = datetime.fromtimestamp(linux_time)
+        formatted_start_time = start_time.strftime('%Y-%m-%dT%H:%M:%S')
```

Comment:
Why do you need to stringify datatetime here? It would be much cleaner if it returned `Dict[str, Tuple[str, datatime]]`

Reply:
Seems the format is different in using upcoming maintenance api vs slurm reservation list. I updated to keep it as it is until we need to use it to make reservation.